src/layers.rs

```diff
@@ -42,42 +43,145 @@ impl LayerTile {
     }
 }
 
-#[derive(Debug, PartialEq, Clone)]
+#[derive(Clone, PartialEq, Debug)]
+pub enum LayerType {
+    TileLayer(TileLayer),
+    ObjectLayer(ObjectLayer),
+    ImageLayer(ImageLayer),
+    // TODO: Support group layers
+}
+
+impl LayerType {
+    pub fn as_tile_layer(&self) -> Option<&TileLayer> {
+        match self {
+            LayerType::TileLayer(x) => Some(x),
+            _ => None,
+        }
+    }
+    pub fn as_tile_layer_mut(&mut self) -> Option<&mut TileLayer> {
+        match self {
+            LayerType::TileLayer(x) => Some(x),
+            _ => None,
+        }
+    }
+
+    pub fn as_object_layer(&self) -> Option<&ObjectLayer> {
+        match self {
+            LayerType::ObjectLayer(x) => Some(x),
+            _ => None,
+        }
+    }
+    pub fn as_object_layer_mut(&mut self) -> Option<&mut ObjectLayer> {
+        match self {
+            LayerType::ObjectLayer(x) => Some(x),
+            _ => None,
+        }
+    }
+
+    pub fn as_image_layer(&self) -> Option<&ImageLayer> {
+        match self {
+            LayerType::ImageLayer(x) => Some(x),
+            _ => None,
+        }
+    }
+    pub fn as_image_layer_mut(&mut self) -> Option<&mut ImageLayer> {
+        match self {
+            LayerType::ImageLayer(x) => Some(x),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Clone, Copy)]
+pub(crate) enum LayerTag {
+    TileLayer,
+    ObjectLayer,
+    ImageLayer,
+}
+
+#[derive(Clone, PartialEq, Debug)]
 pub struct Layer {
     pub name: String,
-    pub width: u32,
-    pub height: u32,
-    pub opacity: f32,
+    pub id: u32,
     pub visible: bool,
     pub offset_x: f32,
     pub offset_y: f32,
-    /// The tiles are arranged in rows. Each tile is a number which can be used
-    ///  to find which tileset it belongs to and can then be rendered.
-    pub tiles: LayerData,
+    pub opacity: f32,
     pub properties: Properties,
-    pub layer_index: u32,
-    /// The ID of the layer, as shown in the editor.
-    /// Layer ID stays the same even if layers are reordered or modified in the editor.
-    pub id: u32,
+    pub ty: LayerType,
```

Comment:
Let's just call this `layer_type`.

Reply:
I was considering a name like that, but it is way too verbose (plus the `layer` part is already implicit from the parent type). It cannot be named `type` either because it's a keyword, so think `ty` works fine. In my mind `LayerType` is a type-specific "extension" of `Layer` so it makes sense to have a short name for it.

Reply:
I think it's really unfortunate that Rust devs decided to make "type" a keyword...

Anyway, I'd still prefer a verbose name over `ty`. I'm using the [same name in Tiled](https://github.com/mapeditor/tiled/blob/master/src/libtiled/layer.h#L85), so it seems rather natural to me.